internal/v1alpha2/store/store.go

```diff
@@ -0,0 +1,193 @@
+		return nil, err
+	}
+	if err := fs.MkdirAll(filepath.Join(baseUnpackPath, bundledeploymentName), 0755); err != nil {
+		return nil, err
+	}
+
+	return &bundledeploymentStore{
+		bundledeploymentName: bundledeploymentName,
+		baseDirectory:        filepath.Join(baseUnpackPath, bundledeploymentName),
+		fs:                   afero.NewBasePathFs(fs, filepath.Join(baseUnpackPath, bundledeploymentName)),
+	}, nil
+}
+
+// Copies contents from a tar reader to the destination on the filesystem
+func (b *bundledeploymentStore) CopyTarArchive(reader *tar.Reader, destination string) error {
+	if reader == nil {
+		return fmt.Errorf("%w for bundle deployment %q", ErrCopyContents, b.GetBundleDeploymentName())
+	}
+
+	dst := filepath.Clean(destination)
+	if dst == "." || dst == "/" {
+		dst = ""
+	}
+
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if header.Typeflag == tar.TypeDir {
+			if err := b.fs.MkdirAll(filepath.Join(dst, header.Name), 0755); err != nil {
+				return fmt.Errorf("%w: %v", ErrCopyContents, err)
+			}
+		} else if header.Typeflag == tar.TypeReg {
+			// If it is a regular file, create the path and copy data.
+			// The header stream is not sorted to go over the directories and then
+			// the files. In case, a file is encountered which does not have a parent
+			// when we try to copy contents from the reader, we would error. So, verify if the
+			// parent exists and then copy contents.
+
+			if err := ensureParentDirExists(b.fs, filepath.Join(dst, header.Name)); err != nil {
+				return fmt.Errorf("%w: %v", ErrCopyContents, err)
+			}
+
+			file, err := b.fs.Create(filepath.Join(dst, header.Name))
+			if err != nil {
+				return fmt.Errorf("%w: %v", ErrCopyContents, err)
+			}
+
+			if _, err := io.Copy(file, reader); err != nil {
+				return fmt.Errorf("%w: %v", ErrCopyContents, err)
+			}
+			file.Close()
+		} else {
+			return fmt.Errorf("%w: unsupported tar entry type %q: %q", ErrCopyContents, header.Name, header.Typeflag)
+		}
+	}
+	return nil
+}
+
+// ensureParentDirExists makes sure that the parent directory for the
+// passed header path exists.
+func ensureParentDirExists(fs afero.Fs, header string) error {
+	parent := filepath.Dir(header)
+
+	// indicates that the file is in cwd
+	if parent == "." || parent == "" {
+		return nil
+	}
+	return fs.MkdirAll(parent, 0755)
+}
+
+func (b *bundledeploymentStore) CopyDirFS(source, destination string, _ afero.Fs) error {
+	return cp.Copy(source, destination, cp.Options{
+		OnDirExists: func(src, dest string) cp.DirExistsAction {
+			return cp.Merge
+		},
+	})
+}
+
+func (b *bundledeploymentStore) GetBundleDeploymentName() string {
+	return b.bundledeploymentName
+}
+
+func (b *bundledeploymentStore) GetBundleDirectory() string {
+	return b.baseDirectory
+}
+
+func (b *bundledeploymentStore) Create(name string) (afero.File, error) {
+	return b.fs.Create(name)
```

Comment:
If you do want to inline everything, why not just embed the `fs`?

Reply:
That's a good point. The only benefit for doing this was creating a chroot like view for the bundle deployment while creating a newBundleDeploymentStore in here: https://github.com/operator-framework/rukpak/pull/743/files#diff-f1022d4c85936c453f8a541cdd88c667367ac932b582d3dadaeb01dd547c9d8bR64. But this could be offloaded in the reconciler itself. 2 additional lines in there, can save a bunch of code here. Will refactor this out.